### PR TITLE
Build and register a block metadata collection in one file for enhanced performance

### PR DESCRIPTION
**Pull request by mreishus**
Take advantage of https://core.trac.wordpress.org/changeset/59132 - "Editor: Allow registering PHP manifest file for block metadata collections for enhanced performance."

* Adds a new `./assets/client/blocks/blocks-json.php` to the built version of the plugin, that compiles all of the block.json files into one php file.
* Calls the new `wp_register_block_metadata_collection()` to register the file, such that subsequent calls to `register_block_from_metadata` don't need to read and parse each individual block.json file.

### To test:

* **Make sure to test with WordPress 6.7 Beta 1 or later.**
  * `function_exists( 'wp_register_block_metadata_collection' )` will return true if this condition is met.
* All blocks should continue to work the same in the editor and in the frontend.
* In `wp-includes/src`, add logging to `register_block_type_from_metadata() near the top:

```php
	if ( $registry_metadata ) {
		$metadata = $registry_metadata;
		error_log("Registering without decoding: $file_or_folder");
	} elseif ( $metadata_file_exists ) {
		$metadata = wp_json_file_decode( $metadata_file, array( 'associative' => true ) );
		error_log("Need to decode: $file_or_folder");
	} else {
		$metadata = array();
	}
```

Check the log and we should see all the woocommerce blocks no longer needing to `wp_json_file_decode()`.

### CODE REVIEW

**Comment by mreishus** [comment]
I believe there's a possible dependency/ordering issue here:
https://github.com/mreishus/woocommerce/blob/a3f4439db072dddbc43f88cc7ec99d8c3b40ac24/plugins/woocommerce-blocks/bin/blocks-json-to-php.js#L20

This is where the build step inside `plugins/woocommerce-blocks` is "reaching into" `plugins/woocommerce` to include a few blocks in the built php file. Depending on what order the clean and build commands are run in, these aren't always included. What I'm looking for to debug this issue is `product-rating`. If that's missing from the built file, then something went wrong.

**Comment by mreishus** [comment]
Explaining the problem in more detail:

### Clean build:
```
pnpm run clean
pnpm run clean:build
pnpm install --frozen-lockfile
pnpm run build
wc -l plugins/woocommerce-blocks/build/blocks-json.php
## See: 4928
```

### More comprehensive build:
```
After running more build commands in the various plugin directories, a `pnpm run build` will eventually create a larger blocks-json.php file.
wc -l plugins/woocommerce-blocks/build/blocks-json.php
## See: 7335
```

### Problem 1: Build Stability

**Expect to see**: Creating a build or zip file will always create a `blocks-json.php` with the same blocks included, not depending on the state of the repo.
**Actually see**: A clean build is behaving differently than a repo that has had the individual plugins built.

### Problem 2: Extra order-summary blocks generated.

Some of the extra blocks generated look to be duplicates, except `-block` is added to the name:


### Problem 3: Some order-confirmation and product blocks missing on a clean build:

These blocks are missing on a clean build (all of these +add-to-cart-form)

**Comment by mreishus** [comment]
For problem 3, let's look at one example closer:

`product-filter-rating`:
  * A "built" repository has `./plugins/woocommerce/assets/client/blocks/product-filter-rating/block.json` created by a build process.
  * A clean repository is missing `./plugins/woocommerce/assets/client/blocks/product-filter-rating/block.json`.
  * Both clean and built repositories have `plugins/woocommerce-blocks/assets/js/blocks/product-filters/inner-blocks/rating-filter/block.json` that have the same content, but are in a different directory name.

The registration process keys off directory name, so it's important that the build has the same directory names that are used when registering.

Note: If we look in `./plugins/woocommerce-blocks/bin/webpack-entries.js` we see some entries with custom directories. For example: 

```
	'product-filter-rating': {
		customDir: 'product-filters/inner-blocks/rating-filter',
		isExperimental: true,
	},
```
https://github.com/mreishus/woocommerce/blob/c4e758bab6607eee372a3bf55dfa2a76782aa8bb/plugins/woocommerce-blocks/bin/webpack-entries.js#L111-L114

This has to be a key part of the confusion.

**Comment by mreishus** [comment]
:heavy_check_mark:  OK, I think I fixed it. 

What's happening is:
* Step 1: _woocommerce-blocks_ plugin: `pnpm run build:project:bundle`: This creates new block.json files inside the build directory.
* Step 2: _woocommerce_ plugin: `pnpm run build:project:copy-assets:blocks`: This copies those block.json files from woocommerce-block's build directory into woocommerce's build directory.

I made my script looking for items in step 2, but I didn't make it dependent upon either step 2 or step 1. So instead, I changed it to look for the items from step 1, and to make it dependent upon step 1.  Phew!

**Comment by mreishus** [comment]
:heavy_check_mark:  Additionally, I removed the duplicate blocks that were being built before the proper name transformations.
After a0c9449:


For example, instead of building the same information under both `cart-order-summary-coupon-form` and `cart-order-summary-coupon-form-block`, it will only be listed under the latter, which is the only one used. In my testing, this reduced the size of the manifest file from ~160KB to ~120KB while keeping all woocommerce blocks to be registered without needing to decode a json file.

**Comment by roykho** [comment]
Closes https://github.com/woocommerce/woocommerce/issues/52125

File: plugins/woocommerce-blocks/bin/blocks-json-to-php.js
Comment (louwie17) [inline plugins/woocommerce-blocks/bin/blocks-json-to-php.js]: It looks like they have [actually released this](https://make.wordpress.org/core/2024/10/17/new-block-type-registration-apis-to-improve-performance-in-wordpress-6-7/) as part of the latest `@wordpress/scripts` version. So we could just update the /scripts version and use `wp-scripts build-blocks-manifest` instead.

File: plugins/woocommerce-blocks/package.json
Comment (louwie17) [inline plugins/woocommerce-blocks/package.json]: We should include the `"dependencyOutputs"` here, this will also auto update the `output` array with the items that got updated.

File: plugins/woocommerce/src/Blocks/BlockTypesController.php
Comment (louwie17) [inline plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: Using `WC_ABSPATH` does seem like the right approach here, but seems to break in my local environment as I have WooCommerce symlinked. I left a question in core around this: https://wordpress.slack.com/archives/C02QB2JS7/p1729539333146749

File: plugins/woocommerce-blocks/package.json
Comment (louwie17) [inline plugins/woocommerce-blocks/package.json:53]: Nice, appreciate integrating it with wireit right away 🎉

File: plugins/woocommerce/src/Blocks/BlockTypesController.php
Comment (louwie17) [reply plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: This may not be a blocker though, given it would just default to the old way if things are symlinked.

**Review by louwie17 — CHANGES_REQUESTED** [review]
@mreishus thanks for getting on this pretty quickly in time for the WP 6.7 release.
It took me a bit to get it working, mostly because the way my local environment is set up ( symlinks ), see the inline comment for more context.

I also left a couple other suggestions, one around using the `wp-scripts build-blocks-manifest` command instead of the `blocks-json-to-php.js` file. I believe this [only got released](https://make.wordpress.org/core/2024/10/17/new-block-type-registration-apis-to-improve-performance-in-wordpress-6-7/#full-example) after you started working on this PR.

File: plugins/woocommerce/src/Blocks/BlockTypesController.php
Comment (louwie17) [reply plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: Realizing this last comment is a bit ambiguous.
I mean that the above issue isn't a blocker and this PR could be merged when the other items are addressed. We can create a follow up PR around this issue if changes on our end are necessary.

File: plugins/woocommerce-blocks/bin/blocks-json-to-php.js
Comment (mreishus) [reply plugins/woocommerce-blocks/bin/blocks-json-to-php.js]: I'm worried about breaking changes when updating to `30.3.0`. I'm not 100% sure it's an issue because the way the monorepo splits up packages and plugins, but checking the [changelog](https://github.com/WordPress/gutenberg/blob/HEAD/packages/scripts/CHANGELOG.md), there are a few breaking changes. I think it's okay because `./woocommerce-blocks` uses React 18 (2022-12-14 breaking changes) and doesn't use `lint-md-js` (2022-05-04), do you know how to test if it's safe to upgrade?

File: plugins/woocommerce-blocks/package.json
Comment (mreishus) [reply plugins/woocommerce-blocks/package.json]: Is this right? 4b48f187ff

File: plugins/woocommerce/src/Blocks/BlockTypesController.php
Comment (mreishus) [reply plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: I see - so it could be failing because it's checking `/Users/dev/projects/woocommerce/assets/client/blocks` against `/srv/http/wordpress/wp-content/plugins` and thinks it's not a specific plugin? That's unfortunate, because some hosts may be linking plugins in production as well.

File: plugins/woocommerce-blocks/bin/blocks-json-to-php.js
Comment (mreishus) [reply plugins/woocommerce-blocks/bin/blocks-json-to-php.js]: I found something else that makes me not want to upgrade: The `28.0.0 (2024-05-31)` breaking change says:
```
Note If you're using @wordpress/scripts for building JS scripts to target WordPress 6.5 or earlier, you should not upgrade to this version and continue using @wordpress/scripts@27.
```
And WC still supports 6.5 per [this document](https://woocommerce.com/document/update-php-wordpress/) and slack convo. p1730824429807539-slack-C03CPM3UXDJ

Comment (mreishus) [reply plugins/woocommerce-blocks/bin/blocks-json-to-php.js]: Ah, I see anything merged to trunk now won't be released until WC 9.5, which will support WP 6.6. p1730828228999269/1730824429.807539-slack-C03CPM3UXDJ

Went ahead and updated it to use my code from wp-scripts.

File: plugins/woocommerce/src/Blocks/BlockTypesController.php
Comment (louwie17) [reply plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: Yeah exactly, and it sounds like from this issue and conversation that they won't change it: https://core.trac.wordpress.org/ticket/62140

We can update it on our end, and instead of relying on `WC_ABSPATH` heavily we actually make use of the plugin directory path, as this should still work.

**Review by louwie17 — APPROVED** [review]
Thanks for the updates @mreishus and updating `@wordpress/scripts`. I think we were also fine updating it anyway as `woocommerce-blocks` doesn't use it for building the JS and still uses a custom webpack version.
So the [breaking change](https://github.com/WordPress/gutenberg/blob/HEAD/packages/scripts/CHANGELOG.md#breaking-changes-2) didn't adhere that you mentioned, but I appreciate you calling that out.

I am approving this, as it tested well both in WP 6.6 and 6.7. It also remained to work well on my local even though the new function didn't work ( it still didn't break anything ). So with those steps in mind it should be fine to approve.
I am going to start a follow up discussion on this, as I don't like that now the plugin technically behaves different from being symlinked versus not, and for testing purposes I think its important it behaves the same way in all environments ( whether symlinked or not ).

File: plugins/woocommerce/src/Blocks/BlockTypesController.php
Comment (senadir) [reply plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: Several managed hosts run Woo symlinked and managed from another path, would this be a problem?

Comment (senadir) [reply plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: `$this->asset_api->package->get_path( 'assets/client/blocks/blocks-json.php' );` Should work in that file, otherwise to make it shorter you can inject the package in bootstrap and use it directly.

Comment (mreishus) [reply plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: ## Symlinking - What happens?

>Several managed hosts run Woo symlinked and managed from another path, would this be a problem?

It means that they won't be able to take advantage of the optimization, but it won't crash or anything. If the paths don't match up, it simply falls back to the old behavior.

## Trying `package->get_path()`

>`$this->asset_api->package->get_path( 'assets/client/blocks/blocks-json.php' );` Should work in that file

This is a fatal error: `Fatal error: Uncaught Error: Cannot access private property Automattic\WooCommerce\Blocks\Assets\Api::$package`.

I was able to find ` $meta_file_path3 = \Automattic\WooCommerce\Blocks\Package::container()->get( \Automattic\WooCommerce\Blocks\Domain\Package::class )->get_path( 'assets/client/blocks/blocks-json.php' );`, but as far as I can tell, this is the exact same as `WC_ABSPATH . 'assets/client/blocks/blocks-json.php';`. For example, when I have `plugins/woocommerce` linked to `/tmp2/woocommerce`, both are `/tmp2/woocommerce/assets/client/blocks/blocks-json.php`.

From trying to pull apart the DI, Package seems to be initialized with the abs path:
```
/tmp2/woocommerce/src/Blocks/Package.php:94 - __construct("11.8.0-dev", "\/tmp2\/woocommerce", {})
```

It might be coming from this `__DIR__` in `src/Blocks/Package.php`:
```
	public static function container( $reset = false ) {
		static $container;
		if (
			! $container instanceof Container
			|| $reset
		) {
			$container = new Container();
			// register Package.
			$container->register(
				NewPackage::class,
				function ( $container ) {
					// leave for automated version bumping.
					$version = '11.8.0-dev';
					return new NewPackage(
						$version,
						dirname( __DIR__, 2 ),
						new FeatureGating()
					);
				}
			);
```

## Changing `WC_ABSPATH` -> `WP_PLUGIN_DIR`

Changing the manifest registration is easy enough:
```
	public function register_block_metadata() {
		$meta_file_path = WP_PLUGIN_DIR . '/woocommerce/assets/client/blocks/blocks-json.php';
		if ( function_exists( 'wp_register_block_metadata_collection' ) && file_exists( $meta_file_path ) ) {
			wp_register_block_metadata_collection(
				WP_PLUGIN_DIR . '/woocommerce/assets/client/blocks/',
				$meta_file_path
			);
		}
	}
```

However, the block registrations also need to be changed to match:
```
	protected function register_block_type() {
        // ..truncated..
		$metadata_path = $this->asset_api->get_block_metadata_path( $this->block_name );
```

This I don't know how to do safely because there are so many layers of OO and DI. The Api.php has:
```
	public function get_block_metadata_path( $block_name, $path = '' ) {
		$path_to_metadata_from_plugin_root = $this->package->get_path( 'assets/client/blocks/' . $path . $block_name . '/block.json' );
```

I think that depends on how the Package class gets instantiated, but I'd rather leave that up to someone else more familiar with the codebase :pray:.

**Comment by mreishus** [comment]
I fixed the conflicts - can I get workflow approval?

**Comment by senadir** [comment]
I approved but there's a conflict still in pnpm lock which you might need to fix locally (nvm use && pnpm install)

**Comment by mreishus** [comment]
I've resolved the conflicts again.

I also had to update some test snapshots which had minor changes after the `npm install`: 
6ae9d895ec44ece2c516de567768692356422f6f.

However, I think some tests are still failing related to upgrading `@wordpress/scripts`. I believe some of the failures are due to `@wordpress/viewport` trying to attach event listeners during test setup. I eventually was able to fix these by adding new calls to `jest.mock()` for `@wordpress/viewport` and `@wordpress/editor`, but I really don't know if that was the right thing to do, and I also was getting a timeout on the `'replaces chosen option when another one is clicked'` test which I don't know how to fix. It also feels like these changes are getting a bit out of scope.

**Impeding progress:**
* Each time someone pushes to trunk changing the lock file, this branch becomes conflicted.
* Whenever I push to this branch, the CI/CD doesn't run unless someone on WooCommerce manually starts the flow.

**I have two ideas for moving this forward:**
* If someone more familiar with the WooCommerce codebase can upgrade plugins/woocommerce-block's @wordpress/scripts 24.6.0 -> 30.4.0 for me in a separate PR, then this PR can be focused on using the new package without having to also handle the dependency update.
* Otherwise, if this PR can go back to using a custom script instead of using the script in @wordpress/scripts (which I also wrote), then the update becomes simpler.

**Comment by joshuatf** [comment]
Hey @mreishus 👋  now that https://github.com/woocommerce/woocommerce/pull/53619 is merged, I think this should be unblocked.  Let me know if you have any more trouble and I'm happy to hop on and review/approve any workflows.

File: plugins/woocommerce/src/Blocks/BlockTypesController.php
Comment (mreishus) [reply plugins/woocommerce/src/Blocks/BlockTypesController.php:174]: I added a bypass for the warning message here: 4a0f4c11b3a4986318bea2df17e399851c040303
This is the same approach we used in JP after the notices were causing alarm with some customers.
I just went through a clean install by building a woocommerce zip, setting WP_DEBUG to be true, and installing woocommerce off a symlinked directory to make sure the bypass was working.

I have also added a comment to https://core.trac.wordpress.org/ticket/62140 but it doesn't seem to be getting much traction.

**Comment by mreishus** [comment]
> Hey @mreishus 👋 now that #53619 is merged, I think this should be unblocked. Let me know if you have any more trouble and I'm happy to hop on and review/approve any workflows.

Thank you! Please kick off the regular checks ( @joshuatf ).

**Review by joshuatf — APPROVED** [review]
Opened https://github.com/woocommerce/woocommerce/pull/53905 to confirm tests are passing since this is a remote branch and not all tests can run correctly.

- ✅ Re-tested this branch and blocks work as expected
- ✅ WC Blocks all use the built blocks file without decoding
- ⚠️  Product editor blocks don't use this registration flow, but that's probably okay since those will transition to DataForm components
- ⚠️ The non-decoded option can't be used with symlinked plugins because of the [plugin path check](https://github.com/WordPress/WordPress/blob/1747403b13a6d20f52082237ab8b8b28d605a4a0/wp-includes/class-wp-block-metadata-registry.php#L112-L120), but this is an issue that will need to be fixed upstream

The two minor issues above are not blockers to merging this PR and things continue to work as expected, so approving this as is.

**Comment by louwie17** [comment]
> ⚠️ The non-decoded option can't be used with symlinked plugins because of the [plugin path check](https://github.com/WordPress/WordPress/blob/1747403b13a6d20f52082237ab8b8b28d605a4a0/wp-includes/class-wp-block-metadata-registry.php#L112-L120), but this is an issue that will need to be fixed upstream

@joshuatf  I wonder if we should explicitly mention this somewhere both internally for awareness and also externally with the next WC release.
While I am not opposed at all having merged this, I could foresee issues cropping up that would not be reproducible in a symlinked environment versus a non-symlinked environment ( or vice versa ), and it could be difficult to track that down without having awareness of this change.

**Pull request merged**
